crypto/mlkem/mlkem-c.inc

```diff
@@ -0,0 +1,1212 @@
+ *
+ * Compresses (lossily) an input |x| mod 3329 into |bits| many bits by grouping
+ * numbers close to each other together. The formula used is
+ * round(2^|bits|/kPrime*x) mod 2^|bits|.
+ * Uses Barrett reduction to achieve constant time. Since we need both the
+ * remainder (for rounding) and the quotient (as the result), we cannot use
+ * |reduce| here, but need to do the Barrett reduction directly.
+ */
+static uint16_t compress(uint16_t x, int bits)
+{
+    uint32_t shifted = (uint32_t)x << bits;
+    uint64_t product = (uint64_t)shifted * kBarrettMultiplier;
+    uint32_t quotient = (uint32_t)(product >> kBarrettShift);
+    uint32_t remainder = shifted - quotient * kPrime;
+
+    /*
+     * Adjust the quotient to round correctly:
+     *   0 <= remainder <= kHalfPrime round to 0
+     *   kHalfPrime < remainder <= kPrime + kHalfPrime round to 1
+     *   kPrime + kHalfPrime < remainder < 2 * kPrime round to 2
+     */
+    assert(remainder < 2u * kPrime);
+    quotient += 1 & constant_time_lt_32(kHalfPrime, remainder);
+    quotient += 1 & constant_time_lt_32(kPrime + kHalfPrime, remainder);
+    return quotient & ((1 << bits) - 1);
+}
+
+/*
+ * FIPS 203, Section 4.2.1, Equation (4.8): Decompress_d.
+
+ * Decompresses |x| by using a close equi-distant representative. The formula
+ * is round(kPrime/2^|bits|*x). Note that 2^|bits| being the divisor allows us
+ * to implement this logic using only bit operations.
+ */
+static uint16_t decompress(uint16_t x, int bits)
+{
+    uint32_t product = (uint32_t)x * kPrime;
+    uint32_t power = 1 << bits;
+    /* This is |product| % power, since |power| is a power of 2. */
+    uint32_t remainder = product & (power - 1);
+    /* This is |product| / power, since |power| is a power of 2. */
+    uint32_t lower = product >> bits;
+
+    /*
+     * The rounding logic works since the first half of numbers mod |power| have a
+     * 0 as first bit, and the second half has a 1 as first bit, since |power| is
+     * a power of 2. As a 12 bit number, |remainder| is always positive, so we
+     * will shift in 0s for a right shift.
+     */
+    return lower + (remainder >> (bits - 1));
+}
+
+static void scalar_compress(scalar *s, int bits)
+{
+    int i;
+
+    for (i = 0; i < DEGREE; i++)
+        s->c[i] = compress(s->c[i], bits);
+}
+
+static void scalar_decompress(scalar *s, int bits)
+{
+    int i;
+
+    for (i = 0; i < DEGREE; i++)
+        s->c[i] = decompress(s->c[i], bits);
+}
+
+static void vector_compress(vector *a, int bits)
+{
+    int i;
+
+    for (i = 0; i < RANK; i++)
+        scalar_compress(&a->v[i], bits);
+}
+
+static void vector_decompress(vector *a, int bits)
+{
+    int i;
+
+    for (i = 0; i < RANK; i++)
+        scalar_decompress(&a->v[i], bits);
+}
+
+int encode_public_key(uint8_t *out, const public_key *pub)
+{
+    vector_encode(out, &pub->t, kLog2Prime);
+    memcpy(out + VECTOR_BYTES, pub->rho, MLKEM_RANDOM_BYTES);
+    return 1;
+}
```

Comment:
This function (under its *variant-specific* ossl_mlkem_... name)  is currently not used outside this file, because the encoded public key is output as part of key generation, so does not typically need to be obtained separately.  The provider is able to return encoded public keys, because they are stored as part of the encoded private key.

So this function could be made `static int`, or kept as-is *just in-case*. Please let me know your thoughts...

Reply:
fwiw, making it static is fine with me if/as the provider code pulls encoded public keys from encoded private keys. I refrained from doing such algorithm-specific logic in provider code such as to keep that as "general" as possible (for algorithms/code that doesn't permit that). But as the low level code here and the provider code really are "an item" now, this approach sounds logical.

Reply:
It can't be kept as is because it doesn't have the `ossl_` prefix we require on all _visible_ symbols (in the `.a` library).

My suggestion would be making it static rather than changing it's name.

Reply:
You've neglected to take into account the macros that give this a per-variant name with an `ossl_mlkem<variant>` prefix. So it correctly exported for now. But I'll make it static and drop the macro.